utils/src/merkle_tree/optimal_merkle_tree.rs

```diff
@@ -276,100 +278,75 @@ where
-
-        // current depth & index (used to compute the hash)
-        // current depth initially == tree depth (or leaves depth)
+    /// Updates parent hashes after modifying a range of leaf nodes.
+    ///
+    /// - `start`: Starting leaf index that was updated.
+    /// - `length`: Number of consecutive leaves that were updated.
+    fn update_hashes(&mut self, start: usize, length: usize) -> Result<()> {
+        // Start at the leaf level
         let mut current_depth = self.depth;
-        let mut current_index = if index % 2 == 0 { index } else { index - 1 };
-        let mut current_index_bak = current_index;
-
-        loop {
-            // Hash 2 values at (current depth, current_index) & (current_depth, current_index + 1)
-            let n_hash = self.hash_couple(current_depth, current_index);
-            // Insert this hash at (parent_depth, parent_index)
-            self.nodes.insert((parent_depth, parent_index), n_hash);
 
-            if parent_depth == 0 {
-                // We just set the root hash of the tree - nothing to do anymore
-                break;
-            }
-            // Incr parent index
-            parent_index += 1;
-            // Incr current index (+2 because we've just hashed current index & current_index + 1)
-            current_index += 2;
-            if parent_index >= parent_max_index {
-                // reset (aka decr depth & reset indexes)
-                parent_depth -= 1;
-                parent_index = parent_index_bak >> 1;
-                parent_index_bak = parent_index;
-                parent_max_index >>= 1;
-                current_depth -= 1;
-                current_index = current_index_bak >> 1;
-                current_index_bak = current_index;
+        // Round down to include the left sibling in the pair (if start is odd)
+        let mut current_index = start & !1;
+
+        // Compute the max index at this level, round up to include the last updated leaf’s right sibling (if start + length is odd)
+        let mut current_index_max = (start + length + 1) & !1;
+
+        // Traverse from the leaf level up to the root
+        while current_depth > 0 {
+            // Compute the parent level (one level above the current)
+            let parent_depth = current_depth - 1;
+
+            // Use parallel processing when the number of pairs exceeds the threshold
+            if current_index_max - current_index >= MIN_PARALLEL_NODES {
+                let updates: Vec<((usize, usize), H::Fr)> = (current_index..current_index_max)
+                    .step_by(2)
+                    .collect::<Vec<_>>()
+                    .into_par_iter()
+                    .map(|index| {
+                        // Hash two child nodes at positions (current_depth, index) and (current_depth, index + 1)
+                        let hash = self.hash_couple(current_depth, index);
+                        // Return the computed parent hash and its position at (parent_depth, index >> 1)
```

Comment:
```suggestion
                        // Return the computed parent hash and its position
```
I think it's clear enough without repeating return value